File: android/app/src/test/java/ai/oriveo/community/core/provider/PresentationContractTest.kt
package ai.oriveo.community.core.provider

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import org.junit.Assert.assertEquals
import org.junit.Test
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

/**
 * Golden tests for the presentation contract.
 *
 * Reads `shared/model-contracts/presentation_fixtures.v1.json`, feeds the fixture's
 * `leafCases` / `containerCases` / `stateCases` into the local pure selector functions
 * ([derivePresentation] / [deriveContainerPresentation] / [deriveStatePresentation]) and
 * asserts that every field of the output matches the fixture's `expectedRender`.
 *
 * These selectors are themselves the single source of truth for the presentation contract on
 * this platform: any rendering work (the Compose provider catalog, the model picker) has to
 * reuse these rules. When an implementation drifts away from the fixture this suite has to go
 * red, so that the drift cannot be committed.
 */
class PresentationContractTest {

    // ── Fixture schema ──

    @Serializable
    private data class FixtureFile(
        val contractVersion: Int,
        val leafCases: List<LeafCase> = emptyList(),
        val containerCases: List<ContainerCase> = emptyList(),
        val stateCases: List<StateCase> = emptyList(),
    )

    @Serializable
    private data class LeafCase(
        val id: String,
        val providerKind: String,
        val model: FixtureModel,
        val expectedRender: PresentationDescriptor,
    )

    @Serializable
    private data class ContainerCase(
        val id: String,
        val providerKind: String,
        val models: List<FixtureModel>,
        val expectedRender: ContainerDescriptor,
    )

    @Serializable
    private data class StateCase(
        val id: String,
        val providerKind: String,
        val metadataSource: String? = null,
        val providerData: FixtureProviderData,
        val manualRetainedModels: List<FixtureManualRetainedModel> = emptyList(),
        val expectedRender: StateDescriptor,
    )

    @Serializable
    private data class FixtureModel(
        val canonicalModelId: String,
        val displayName: String? = null,
        val vendorKey: String? = null,
        val vendorName: String? = null,
        val contextLength: Int? = null,
        val pricing: FixturePricing? = null,
        val pricingStatus: String,
        val capabilities: List<String> = emptyList(),
        val profiles: FixtureProfiles? = null,
        val uiHints: FixtureUiHints,
    )

    @Serializable
    private data class FixturePricing(
        val promptPerMToken: Double? = null,
        val completionPerMToken: Double? = null,
        val cachedInputPerMToken: Double? = null,
    )

    @Serializable
    private data class FixtureProfiles(
        val reasoning: String? = null,
        val webSearch: String? = null,
        val imageGen: String? = null,
    )

    @Serializable
    private data class FixtureUiHints(
        val groupKey: String? = null,
        val groupName: String? = null,
        val rank: Int? = null,
        val recommended: Boolean = false,
        val badgeOrder: List<String> = emptyList(),
    )

    @Serializable
    private data class FixtureProviderData(
        val displayName: String,
        val defaultModelId: String? = null,
        val validationModelId: String? = null,
        val resolveMap: Map<String, String> = emptyMap(),
        val models: Map<String, FixtureProviderModel> = emptyMap(),
    )

    /**
     * The model shape inside stateCases is simpler than in leafCases: a subset of the fields.
     * Only presence matters here, so everything is optional.
     */
    @Serializable
    private data class FixtureProviderModel(
        val canonicalModelId: String? = null,
        val displayName: String? = null,
        val pricingStatus: String? = null,
        val capabilities: List<String> = emptyList(),
        val uiHints: FixtureUiHints? = null,
    )

    @Serializable
    private data class FixtureManualRetainedModel(
        val modelId: String,
        val displayName: String? = null,
    )

    // Descriptors: the source-of-truth shape that expectedRender is compared against.

    @Serializable
    private data class PresentationDescriptor(
        val showsVendorSubtitle: Boolean,
        val vendorSubtitleText: String? = null,
        val showsRecommendedBadge: Boolean,
        val capabilityBadges: List<String>,
        val showsCachedPricing: Boolean,
        val pricingBranch: String,
        val showsReasoningPicker: Boolean,
        val showsWebSearchPicker: Boolean,
        val showsImageGenPicker: Boolean,
    )

    @Serializable
    private data class ContainerDescriptor(
        val groupHeaders: List<String>,
        val modelOrderWithinGroups: Map<String, List<String>>,
    )

    @Serializable
    private data class StateDescriptor(
        val showsEmptyStateCopy: Boolean,
        val showsRetryAction: Boolean,
        val showsCatalogList: Boolean,
        val showsOfflineBanner: Boolean,
        val showsManualRetainedSection: Boolean,
        val manualRetainedHeaderKey: String? = null,
    )

    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    private val fixture: FixtureFile by lazy {
        val path = fixtureCandidates()
            .firstOrNull { Files.exists(it) }
            ?.normalize()
            ?: error("presentation_fixtures.v1.json not found in any candidate path")
        json.decodeFromString(path.toFile().readText())
    }

    // ── Tests ──

    @Test
    fun `fixture loads with expected contract version`() {
        assertEquals(1, fixture.contractVersion)
    }

    @Test
    fun `leaf cases stay aligned`() {
        fixture.leafCases.forEach { case ->
            val actual = derivePresentation(case.model)
            assertEquals("leaf case `${case.id}`", case.expectedRender, actual)
        }
    }

    @Test
    fun `container cases stay aligned`() {
        fixture.containerCases.forEach { case ->
            val actual = deriveContainerPresentation(case.providerKind, case.models)
            assertEquals("container case `${case.id}`", case.expectedRender, actual)
        }
    }

    @Test
    fun `state cases stay aligned`() {
        fixture.stateCases.forEach { case ->
            val actual = deriveStatePresentation(
                metadataSource = case.metadataSource,
                providerData = case.providerData,
                manualRetainedModels = case.manualRetainedModels,
            )
            assertEquals("state case `${case.id}`", case.expectedRender, actual)
        }
    }

    // Selectors: the contract implementation. Pure functions taking fixture fields in and
    // returning a descriptor.

    /**
     * The rendering contract for a single model card. The rules follow the field-to-UI mapping
     * table in `00-metadata-contract-spec.md` exactly:
     *
     * - vendor subtitle  <= vendorKey and vendorName are both present and non-blank after trim
     * - recommended badge <= uiHints.recommended == true
     * - capability badges <= uiHints.badgeOrder, order preserved, with "text" dropped
     * - pricing branch    <= pricingStatus (priced / free / unknown); showsCachedPricing only
     *                        applies to priced
     * - selectors         <= whether profiles.{reasoning,webSearch,imageGen} are non-empty,
     *                        which is deliberately decoupled from capabilities
     */
    private fun derivePresentation(model: FixtureModel): PresentationDescriptor {
        val vendorKey = model.vendorKey?.trim().orEmpty()
        val vendorName = model.vendorName?.trim().orEmpty()
        val showsVendor = vendorKey.isNotEmpty() && vendorName.isNotEmpty()

        val badges = model.uiHints.badgeOrder.filter { it != "text" }

        val pricingBranch = when (model.pricingStatus) {
            "priced", "free", "unknown" -> model.pricingStatus
            else -> error("unknown pricingStatus `${model.pricingStatus}` in fixture")
        }

        val reasoningProfile = model.profiles?.reasoning?.takeIf { it.isNotBlank() }
        val webSearchProfile = model.profiles?.webSearch?.takeIf { it.isNotBlank() }
        val imageGenProfile = model.profiles?.imageGen?.takeIf { it.isNotBlank() }

        return PresentationDescriptor(
            showsVendorSubtitle = showsVendor,
            vendorSubtitleText = if (showsVendor) vendorName else null,
            showsRecommendedBadge = model.uiHints.recommended,
            capabilityBadges = badges,
            showsCachedPricing = pricingBranch == "priced",
            pricingBranch = pricingBranch,
            showsReasoningPicker = reasoningProfile != null,
            showsWebSearchPicker = webSearchProfile != null,
            showsImageGenPicker = imageGenProfile != null,
        )
    }

    /**
     * The rendering contract for the container, meaning the catalog list.
     *
     * - **Aggregator providers** (OpenRouter, SiliconFlow): grouped by `vendorName`; group
     *   headers ordered by first appearance of the vendor; within a vendor, descending
     *   `uiHints.rank`.
     * - **Direct providers** (everything else): grouped by `uiHints.groupName`; group headers
     *   ordered by each group's highest rank, descending; within a group, descending rank.
     */
    private fun deriveContainerPresentation(
        providerKind: String,
        models: List<FixtureModel>,
    ): ContainerDescriptor {
        val isAggregator = providerKind in AGGREGATOR_PROVIDER_KINDS

        return if (isAggregator) {
            val groups = linkedMapOf<String, MutableList<FixtureModel>>()
            models.forEach { model ->
                val header = model.vendorName.orEmpty().ifEmpty { "Unknown" }
                groups.getOrPut(header) { mutableListOf() }.add(model)
            }
            val headers = groups.keys.toList()
            val orderWithin = groups.mapValues { (_, vendorModels) ->
                vendorModels
                    .sortedByDescending { it.uiHints.rank ?: 0 }
                    .map { it.canonicalModelId }
            }
            ContainerDescriptor(groupHeaders = headers, modelOrderWithinGroups = orderWithin)
        } else {
            val groups = linkedMapOf<String, MutableList<FixtureModel>>()
            models.forEach { model ->
                val header = model.uiHints.groupName.orEmpty().ifEmpty { "Other" }
                groups.getOrPut(header) { mutableListOf() }.add(model)
            }
            val headers = groups.entries
                .sortedByDescending { (_, groupModels) ->
                    groupModels.maxOf { it.uiHints.rank ?: 0 }
                }
                .map { it.key }
            val orderWithin = groups.mapValues { (_, groupModels) ->
                groupModels
                    .sortedByDescending { it.uiHints.rank ?: 0 }
                    .map { it.canonicalModelId }
            }
            ContainerDescriptor(groupHeaders = headers, modelOrderWithinGroups = orderWithin)
        }
    }

    /**
     * The rendering contract for container-level state.
     *
     * - providerData.models empty     -> show the empty state plus a retry button, and no
     *                                    catalog list
     * - providerData.models non-empty -> show the catalog list
     * - metadataSource == "cachedOffline" -> show the offline banner at the top
     * - manualRetainedModels non-empty -> show the Manual-Retained group as its own section
     */
    private fun deriveStatePresentation(
        metadataSource: String?,
        providerData: FixtureProviderData,
        manualRetainedModels: List<FixtureManualRetainedModel>,
    ): StateDescriptor {
        val hasModels = providerData.models.isNotEmpty()
        val hasManualRetained = manualRetainedModels.isNotEmpty()
        val isOffline = metadataSource == "cachedOffline"

        return StateDescriptor(
            showsEmptyStateCopy = !hasModels,
            showsRetryAction = !hasModels,
            showsCatalogList = hasModels,
            showsOfflineBanner = isOffline,
            showsManualRetainedSection = hasManualRetained,
            manualRetainedHeaderKey =
                if (hasManualRetained) MANUAL_RETAINED_HEADER_KEY else null,
        )
    }

    // ── Helpers ──

    private fun fixtureCandidates(): List<Path> {
        val cwd = Paths.get("").toAbsolutePath().normalize()
        return generateSequence(cwd) { current -> current.parent }
            .take(8)
            .map { current ->
                current.resolve("shared/model-contracts/presentation_fixtures.v1.json")
            }
            .toList()
    }

    private companion object {
        /** Matches the providerKind strings used in the fixture, that is the raw values. */
        val AGGREGATOR_PROVIDER_KINDS = setOf("openRouter", "siliconFlow")

        const val MANUAL_RETAINED_HEADER_KEY = "providers.catalog.manualRetainedHeader"
    }
}

File: web/apps/app/lib/core/__tests__/presentation-contract.test.ts
/**
 * Presentation contract golden tests.
 *
 * Shared fixture: model-contracts/presentation_fixtures.v1.json.
 * Every client renders model cards and the provider catalog through one shared set of
 * components, so the structural elements derived from the same fixture must match:
 *   - whether the vendor subtitle is shown
 *   - whether the recommended badge is shown
 *   - capability badge order and count
 *   - pricing branch (priced / free / unknown)
 *   - whether the reasoning / web search / image generation pickers appear
 *   - container level: group header count and order, empty state, offline banner,
 *     Manual-Retained section
 *
 * Strategy: rather than mounting the whole React tree (expensive, needs jsdom + RTL, and
 * ModelCard is widely coupled), pure derive*Presentation() functions take the fixture
 * metadata fields and return a PresentationDescriptor whose shape matches the fixture
 * expectedRender. Those functions are themselves part of the contract: they define how
 * metadata fields map to UI structure assertions.
 */
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { describe, expect, it } from 'vitest';

// Fixture types

type PricingStatus = 'priced' | 'free' | 'unknown';

interface FixturePricing {
  promptPerMToken: number | null;
  completionPerMToken: number | null;
  cachedInputPerMToken: number | null;
}

interface FixtureProfiles {
  reasoning: string | null;
  webSearch: string | null;
  imageGen: string | null;
}

interface FixtureUIHints {
  groupKey?: string;
  groupName?: string;
  rank: number;
  recommended: boolean;
  badgeOrder: string[];
}

interface FixtureModel {
  canonicalModelId: string;
  displayName: string;
  vendorKey: string | null;
  vendorName: string | null;
  contextLength?: number | null;
  pricing?: FixturePricing | null;
  pricingStatus: PricingStatus;
  capabilities: string[];
  profiles?: FixtureProfiles;
  uiHints: FixtureUIHints;
}

interface LeafCase {
  id: string;
  providerKind: string;
  model: FixtureModel;
  expectedRender: ExpectedLeafRender;
}

interface ContainerCase {
  id: string;
  providerKind: string;
  models: FixtureModel[];
  expectedRender: ExpectedContainerRender;
}

interface FixtureProviderData {
  displayName: string;
  defaultModelId?: string | null;
  validationModelId?: string | null;
  resolveMap?: Record<string, string>;
  models: Record<string, FixtureModel>;
}

interface ManualRetainedRef {
  modelId: string;
  displayName: string;
}

interface StateCase {
  id: string;
  providerKind: string;
  metadataSource?: 'freshNetwork' | 'cachedOffline';
  providerData: FixtureProviderData;
  manualRetainedModels?: ManualRetainedRef[];
  expectedRender: ExpectedStateRender;
}

interface ExpectedLeafRender {
  showsVendorSubtitle: boolean;
  vendorSubtitleText: string | null;
  showsRecommendedBadge: boolean;
  capabilityBadges: string[];
  showsCachedPricing: boolean;
  pricingBranch: PricingStatus;
  showsReasoningPicker: boolean;
  showsWebSearchPicker: boolean;
  showsImageGenPicker: boolean;
}

interface ExpectedContainerRender {
  groupHeaders: string[];
  modelOrderWithinGroups: Record<string, string[]>;
}

interface ExpectedStateRender {
  showsEmptyStateCopy: boolean;
  showsRetryAction: boolean;
  showsCatalogList: boolean;
  showsOfflineBanner: boolean;
  showsManualRetainedSection: boolean;
  manualRetainedHeaderKey?: string;
}

interface FixtureFile {
  contractVersion: number;
  leafCases: LeafCase[];
  containerCases: ContainerCase[];
  stateCases: StateCase[];
}

const fixturePath = path.resolve(
  process.cwd(),
  '../../../shared/model-contracts/presentation_fixtures.v1.json',
);
const fixture = JSON.parse(readFileSync(fixturePath, 'utf8')) as FixtureFile;

// Presentation Selector
//
// These three pure functions are the contract for how the shared rendering components
// interpret metadata fields; equivalent fixtures must produce the same descriptor everywhere.

type PresentationDescriptor = ExpectedLeafRender;
type ContainerDescriptor = ExpectedContainerRender;
type StateDescriptor = ExpectedStateRender;

const AGGREGATOR_PROVIDER_KINDS = new Set(['openRouter', 'siliconFlow']);

/**
 * Leaf level: map a single model to a PresentationDescriptor.
 *
 *   - vendor subtitle: shown only when both vendorKey and vendorName are non-empty,
 *     otherwise the whole slot is hidden
 *   - recommended badge: uiHints.recommended === true
 *   - capability badges: badgeOrder order, minus 'text', minus anything absent from capabilities
 *   - pricing branch: read straight from pricingStatus; only priced shows the cache pricing row
 *   - reasoning/webSearch/imageGen pickers: shown only when the matching profiles field is
 *     non-null (capabilities may list reasoning while profiles.reasoning is null, and then no
 *     picker may be shown)
 */
function derivePresentation(model: FixtureModel): PresentationDescriptor {
  const showsVendorSubtitle = Boolean(model.vendorKey && model.vendorName);
  const vendorSubtitleText = showsVendorSubtitle ? (model.vendorName as string) : null;

  const capabilitySet = new Set(model.capabilities);
  const capabilityBadges = (model.uiHints.badgeOrder ?? []).filter(
    (cap) => cap !== 'text' && capabilitySet.has(cap),
  );

  const pricingBranch = model.pricingStatus;
  const showsCachedPricing = pricingBranch === 'priced';

  const profiles = model.profiles ?? { reasoning: null, webSearch: null, imageGen: null };

  return {
    showsVendorSubtitle,
    vendorSubtitleText,
    showsRecommendedBadge: model.uiHints.recommended === true,
    capabilityBadges,
    showsCachedPricing,
    pricingBranch,
    showsReasoningPicker: profiles.reasoning != null,
    showsWebSearchPicker: profiles.webSearch != null,
    showsImageGenPicker: profiles.imageGen != null,
  };
}

/**
 * Container level: map a provider model list to a ContainerDescriptor.
 *
 * Rules:
 *   - aggregator providers (openRouter / siliconFlow): group by vendorName, order group headers
 *     by first appearance of the vendor, sort within a vendor by uiHints.rank descending
 *   - direct providers: group by uiHints.groupName, order group headers by the highest rank in
 *     each group descending, sort within a group by rank descending
 *
 * Tie-breaker: equal ranks sort by canonicalModelId ASCII ascending.
 */
function deriveContainerPresentation(
  providerKind: string,
  models: FixtureModel[],
): ContainerDescriptor {
  const isAggregator = AGGREGATOR_PROVIDER_KINDS.has(providerKind);

  // 1) Group: aggregators by vendorName, direct providers by groupName
  const groupOrder: string[] = [];
  const buckets = new Map<string, FixtureModel[]>();

  for (const model of models) {
    const groupName = isAggregator
      ? (model.vendorName ?? '')
      : (model.uiHints.groupName ?? '');
    if (!buckets.has(groupName)) {
      buckets.set(groupName, []);
      groupOrder.push(groupName);
    }
    buckets.get(groupName)!.push(model);
  }

  // 2) Within a group: rank descending, tie-broken by canonicalModelId ASCII ascending
  const sortedBuckets = new Map<string, FixtureModel[]>();
  for (const [name, list] of buckets) {
    const sorted = [...list].sort((a, b) => {
      const rankDiff = (b.uiHints.rank ?? 0) - (a.uiHints.rank ?? 0);
      if (rankDiff !== 0) return rankDiff;
      return a.canonicalModelId.localeCompare(b.canonicalModelId);
    });
    sortedBuckets.set(name, sorted);
  }

  // 3) Group header order
  let orderedHeaders: string[];
  if (isAggregator) {
    // Aggregator: keep the first-appearance order of vendors, so groupOrder stays as is
    orderedHeaders = groupOrder;
  } else {
    // Direct: highest rank per group descending; equal ranks fall back to group name ASCII ascending for stability
    orderedHeaders = [...groupOrder].sort((a, b) => {
      const maxA = Math.max(...sortedBuckets.get(a)!.map((m) => m.uiHints.rank ?? 0));
      const maxB = Math.max(...sortedBuckets.get(b)!.map((m) => m.uiHints.rank ?? 0));
      if (maxA !== maxB) return maxB - maxA;
      return a.localeCompare(b);
    });
  }

  const modelOrderWithinGroups: Record<string, string[]> = {};
  for (const header of orderedHeaders) {
    modelOrderWithinGroups[header] = sortedBuckets.get(header)!.map((m) => m.canonicalModelId);
  }

  return {
    groupHeaders: orderedHeaders,
    modelOrderWithinGroups,
  };
}

/**
 * State level: map provider data, metadata source and the manual-retained list onto the
 * visibility of the empty state, retry, list, offline banner and Manual-Retained section.
 */
function deriveStatePresentation(input: {
  metadataSource?: 'freshNetwork' | 'cachedOffline';
  providerData: FixtureProviderData;
  manualRetainedModels?: ManualRetainedRef[];
}): StateDescriptor {
  const modelCount = Object.keys(input.providerData.models ?? {}).length;
  const showsCatalogList = modelCount > 0;
  const showsEmptyStateCopy = !showsCatalogList;
  const showsRetryAction = showsEmptyStateCopy;

  const showsOfflineBanner = input.metadataSource === 'cachedOffline';

  const manualCount = input.manualRetainedModels?.length ?? 0;
  const showsManualRetainedSection = manualCount > 0;

  const descriptor: StateDescriptor = {
    showsEmptyStateCopy,
    showsRetryAction,
    showsCatalogList,
    showsOfflineBanner,
    showsManualRetainedSection,
  };
  if (showsManualRetainedSection) {
    descriptor.manualRetainedHeaderKey = 'providers.catalog.manualRetainedHeader';
  }
  return descriptor;
}

// Tests

describe('presentation contract — leaf', () => {
  for (const leafCase of fixture.leafCases) {
    it(`${leafCase.id} — derivePresentation matches expectedRender`, () => {
      const actual = derivePresentation(leafCase.model);
      // expectedRender carries a $comment debug field, so compare only the contract fields
      const expected: ExpectedLeafRender = {
        showsVendorSubtitle: leafCase.expectedRender.showsVendorSubtitle,
        vendorSubtitleText: leafCase.expectedRender.vendorSubtitleText,
        showsRecommendedBadge: leafCase.expectedRender.showsRecommendedBadge,
        capabilityBadges: leafCase.expectedRender.capabilityBadges,
        showsCachedPricing: leafCase.expectedRender.showsCachedPricing,
        pricingBranch: leafCase.expectedRender.pricingBranch,
        showsReasoningPicker: leafCase.expectedRender.showsReasoningPicker,
        showsWebSearchPicker: leafCase.expectedRender.showsWebSearchPicker,
        showsImageGenPicker: leafCase.expectedRender.showsImageGenPicker,
      };
      expect(actual).toEqual(expected);
    });
  }
});

describe('presentation contract — container', () => {
  for (const containerCase of fixture.containerCases) {
    it(`${containerCase.id} — deriveContainerPresentation matches expectedRender`, () => {
      const actual = deriveContainerPresentation(
        containerCase.providerKind,
        containerCase.models,
      );
      const expected: ExpectedContainerRender = {
        groupHeaders: containerCase.expectedRender.groupHeaders,
        modelOrderWithinGroups: containerCase.expectedRender.modelOrderWithinGroups,
      };
      expect(actual).toEqual(expected);
    });
  }
});

describe('presentation contract — state', () => {
  for (const stateCase of fixture.stateCases) {
    it(`${stateCase.id} — deriveStatePresentation matches expectedRender`, () => {
      const actual = deriveStatePresentation({
        metadataSource: stateCase.metadataSource,
        providerData: stateCase.providerData,
        manualRetainedModels: stateCase.manualRetainedModels,
      });
      const expected: ExpectedStateRender = {
        showsEmptyStateCopy: stateCase.expectedRender.showsEmptyStateCopy,
        showsRetryAction: stateCase.expectedRender.showsRetryAction,
        showsCatalogList: stateCase.expectedRender.showsCatalogList,
        showsOfflineBanner: stateCase.expectedRender.showsOfflineBanner,
        showsManualRetainedSection: stateCase.expectedRender.showsManualRetainedSection,
      };
      if (stateCase.expectedRender.showsManualRetainedSection) {
        expected.manualRetainedHeaderKey = stateCase.expectedRender.manualRetainedHeaderKey;
      }
      expect(actual).toEqual(expected);
    });
  }
});
